Cython and Python3: string conversion

Python2 uses ASCII strings by default (`'ü'` returns `'\xc3\xbc'`) and UTF8 strings when explicitly specified (`u'ü'` returns `u'\xfc'`). In both cases the `print()` function is able to correctly display the accentuated character. Python3 uses UTF8 strings by default (`'ü'` returns `'ü'`).

Runtime error messages from the core are all ASCII, so they can be displayed natively in Python2 but in Python3 there is a character type mismatch, so Python3 stores these exception strings as bytestrings (`b'P3M...'`), which the `print()` function cannot display as a regular string. For example, newline characters don't create a new line:
```
  2/124 Test #103: drude ...................................***Failed    4.51 sec
b'P3M tune parameters: Accuracy goal = 1.00000e-04 prefactor = 1.38936e+03\nSystem: box_l = 5.00000e+01 # charged part = 4 Sum[q_i^2] = 8.58771e-01\nfixed mesh 18 18 18\nfixed cao 5\nmesh cao r_cut_iL     alpha_L      err          rs_err     ks_err     time [ms]\n18   5   4.71820e-01 6.42609e+00 9.98037e-05 7.071e-05 7.043e-05 0.41    \n\nresulting parameters: mesh: (18 18 18), cao: 5, r_cut_iL: 4.7182e-01,\n                      alpha_L: 6.4261e+00, accuracy: 9.9804e-05, time: 0.41\n'
ERROR: bond broken between particles 6 and 7. Distance vector: 2.77812 -7.05504 -20.1874
ERROR: bond broken between particles 6 and 7. Distance vector: 2.77812 -7.05504 -20.1874
```

This is maybe related to the following error that prevents sphinx from building the documentation when compiling espresso with python3:
```
Warning, treated as error:
autodoc: failed to import module u'espressomd.__cinit__'; the following exception was raised:
/work/jgrad/espresso-fork-PR/build/src/python/espressomd/_init.so: undefined symbol: PyUnicode_InternFromString
doc/sphinx/CMakeFiles/sphinx.dir/build.make:57: recipe for target 'doc/sphinx/CMakeFiles/sphinx' failed
make[3]: *** [doc/sphinx/CMakeFiles/sphinx] Error 2
CMakeFiles/Makefile2:778: recipe for target 'doc/sphinx/CMakeFiles/sphinx.dir/all' failed
make[2]: *** [doc/sphinx/CMakeFiles/sphinx.dir/all] Error 2
CMakeFiles/Makefile2:785: recipe for target 'doc/sphinx/CMakeFiles/sphinx.dir/rule' failed
make[1]: *** [doc/sphinx/CMakeFiles/sphinx.dir/rule] Error 2
Makefile:383: recipe for target 'sphinx' failed
make: *** [sphinx] Error 2
```

--- a/src/python/espressomd/electrostatics.pxd
+++ b/src/python/espressomd/electrostatics.pxd
@@ -23,7 +23,7 @@ include "myconfig.pxi"
 from espressomd.system cimport *
 cimport numpy as np
 from espressomd.utils cimport *
-from espressomd.utils import is_valid_type
+from espressomd.utils import is_valid_type, to_str
 
 cdef extern from "SystemInterface.hpp":
     cdef cppclass SystemInterface:
@@ -135,7 +135,7 @@ IF ELECTROSTATICS:
             response = p3m_adaptive_tune(& log)
             handle_errors("Error in p3m_adaptive_tune")
             if log.strip():
-                print(log)
+                print(to_str(log))
             return response
 
         cdef inline python_p3m_set_params(p_r_cut, p_mesh, p_cao, p_alpha, p_accuracy):
@@ -224,9 +224,9 @@ IF ELECTROSTATICS:
         if MMM1D_sanity_checks() == 1:
             handle_errors(
                 "MMM1D Sanity check failed: wrong periodicity or wrong cellsystem, PRTFM")
-        resp = mmm1d_tune( & log)
+        resp = mmm1d_tune(& log)
         if resp:
-            print(log)
+            print(to_str(log))
         return resp
 
 IF ELECTROSTATICS:

--- a/src/python/espressomd/electrostatics.pyx
+++ b/src/python/espressomd/electrostatics.pyx
@@ -28,7 +28,7 @@ IF SCAFACOS == 1:
     from .scafacos import ScafacosConnector
     from . cimport scafacos
 from espressomd.utils cimport handle_errors
-from espressomd.utils import is_valid_type
+from espressomd.utils import is_valid_type, to_str
 from . cimport checks
 from .c_analyze cimport partCfg, PartCfg
 from .particle_data cimport particle
@@ -66,7 +66,7 @@ IF ELECTROSTATICS == 1:
 
         def _deactivate_method(self):
             deactivate_method()
-            handle_errors("Coulom method deactivation")
+            handle_errors("Coulomb method deactivation")
 
         def tune(self, **tune_params_subset):
             if tune_params_subset is not None:

--- a/src/python/espressomd/magnetostatics.pyx
+++ b/src/python/espressomd/magnetostatics.pyx
@@ -26,7 +26,7 @@ IF SCAFACOS == 1:
     from . cimport scafacos
 
 from espressomd.utils cimport handle_errors
-from espressomd.utils import is_valid_type
+from espressomd.utils import is_valid_type, to_str
 
 IF DIPOLES == 1:
     cdef class MagnetostaticInteraction(Actor):
@@ -173,7 +173,7 @@ IF DP3M == 1:
             if resp:
                 raise Exception(
                     "failed to tune dipolar P3M parameters to required accuracy")
-            print(log)
+            print(to_str(log))
             self._params.update(self._get_params_from_es_core())
 
         def _activate_method(self):

--- a/src/python/espressomd/utils.pyx
+++ b/src/python/espressomd/utils.pyx
@@ -61,7 +61,7 @@ cpdef check_type_or_throw_except(x, n, t, msg):
     checking is done on the elements, and all elements are checked. Integers
     are accepted when a float was asked for.
 
-     """
+    """
     # Check whether x is an array/list/tuple or a single value
     if n > 1:
         if hasattr(x, "__getitem__"):
@@ -254,13 +254,15 @@ cpdef handle_errors(msg):
 
     """
     errors = mpi_gather_runtime_errors()
+    # print all errors and warnings
     for err in errors:
         err.print()
 
+    # raise an exception with the first error
     for err in errors:
-    # Cast because cython does not support typed enums completely
-        if < int > err.level() == <int > ERROR:
-            raise Exception("{}: {}".format(msg, err.format()))
+        # Cast because cython does not support typed enums completely
+        if < int > err.level() == < int > ERROR:
+            raise Exception("{}: {}".format(msg, to_str(err.format())))
 
 
 def nesting_level(obj):

